src/libsystemd-network/ndisc-protocol.c

```diff
@@ -1,7 +1,45 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
+#include <netinet/icmp6.h>
+
 #include "ndisc-protocol.h"
 
+int ndisc_option_parse(
+                ICMP6Packet *p,
+                size_t offset,
+                uint8_t *ret_type,
+                size_t *ret_len,
+                const uint8_t **ret_opt) {
+
+        assert(p);
+
+        if (offset == p->raw_size)
+                return -ESPIPE; /* end of the packet */
+
+        if (offset > p->raw_size)
+                return -EBADMSG;
+
+        if (p->raw_size - offset < sizeof(struct nd_opt_hdr))
+                return -EBADMSG;
+
```

Comment:
btw, might be nice to add:

```c
assert_cc(alignof(struct nd_opt_hdr) == 1);
```

just to make clear that the following cast is safe regarding alignment.

Reply:
I will add that in a later PR.

Reply:
-> #31555